### ISSUE TITLE
Figure out what it means to assign a bug

### ISSUE THREAD

**Issue by johngruen** [mozilla/fxa-content-server#2549]

**Comment by TDA** [comment]
#BugCeption

**Comment by eoger** [comment]
I like @vladikoff workflow : assign yourself on a buy you intend to work on, then if after two weeks if you didn't have the time to work on it unassign yourself.
The problem is this "timeout" is not automatic and we often forget to unassign ourselves. For example, when I was contributing from home I found interesting issues I wanted to work on, but didn't because someone was assigned to it and I didn't want to "steal" it.

TLDR: I like the "I'm gonna work on it" definition of the assign button, but it should have a "timeout".

**Comment by ckarlof** [comment]
> TLDR: I like the "I'm gonna work on it" definition of the assign button, but it should have a "timeout".

I comfortable with the flexibility in this. I think the next step is to make a PR for the contributing document explaining this policy, what the timeout period is, and if there hasn't been any activity after the timeout period, what some who wants to take the issue should do. @eoger, what to take a crack at it?

**Comment by rfk** [comment]
We could also semi-automate this, e.g. have a bot that periodically checks for timed-out issues and either takes some action, or prompts the owner to take some action.

**Comment by pdehaan** [comment]
Auto-close all bugs where that hasn't been an update/comment in 6 weeks.

**Comment by vladikoff** [comment]
> Auto-close all bugs where that hasn't been an update/comment in 6 weeks.

-1 :scream:

**Comment by shane-tomlinson** [comment]
My own thinking closely mirrors @vladikoff's. I like to assign myself bugs, it's a note to myself "hey, I'm on the hook for this, to at least investigate." If I take further action, it's usually done within 2 weeks. If not, then I start to feel guilty and set the issue free so to speak.

I tag issues with my face if it's 1. a bug I know I caused, or 2. something related to what I have done that needs to be investigated. If the issue does not meet one of those two criteria, I leave it be.

> We could also semi-automate this, e.g. have a bot that periodically checks for timed-out issues and either takes some action, or prompts the owner to take some action.

This would be awesome.

> Auto-close all bugs where that hasn't been an update/comment in 6 weeks.

This makes me quake, though it probably shouldn't.

**Comment by pdehaan** [comment]
>> Auto-close all bugs where that hasn't been an update/comment in 6 weeks.

> This makes me quake, though it probably shouldn't.

Oh, I wasn't serious. That concept completely terrifies me as so many issues would fall through a pretty huge crack. Forcing people to re-file bugs every 6 weeks/months isn't a great system.

**Comment by eoger** [comment]
Here's what I did this afternoon : https://github.com/eoger/issues-bouncer.
Currently it comments the issue and un-assigns people at the same time when 2 weeks have passed.

**Comment by eoger** [comment]
I hacked a little script to check for assigned issues for more than 2 weeks without PRs: 
https://github.com/eoger/issues-bouncer
Currently it un-assigns people and comments on the issue.
Would love some feedback! :+1:

**Comment by rfk** [comment]
Great stuff @eoger, I think we should do more of this kind of automation.

The only trick is to make sure it doesn't generate bad experiences for contributors.  For a small example of automation going off the rails, see gitcop's IMHO-not-that-helpful contributions in this PR:  https://github.com/mozilla/fxa-oauth-server/pull/271

If you and others are interested in hacking on some github automation, I suggest we set up a shared fxa-* repo where we can collaborate.  Thoughts?

**Comment by zaach** [comment]
Awesome @eoger!

Yeah, I think assigning yourself as a way to manage your TODO list is useful. Combined with the "in progress" state, you'll know who's actively working on something vs. who intends to work on it. If you really want to pick up an issue and it's not "in progress", you could always comment in the issue.

**Comment by pdehaan** [comment]
@eoger: bonus points if we call it **pr-cop** and include this image with each automated message:
![](http://i.kinja-img.com/gawker-media/image/upload/s--6ab05BB5--/c_fit,fl_progressive,q_80,w_636/19g09bluaf9pupng.png)

---

But yeah, maybe we pair the **issues-bouncer** with a special GitHub tag where we say "issues-bouncer, ignore this bug and don't auto unassign me". Should be pretty simple to loop over the GitHub issues and ignore anything tagged with **issues-bouncer-ignore**. Or not, whatever.

**Comment by issues-bouncer** [comment]
Hey @johngruen, you have been assigned on this issue for more than 14 days!
Maybe you should consider un-assigning yourself and let other people try to solve this issue.

**Comment by eoger** [comment]
First round of reprimands sent :), feel free to post some feedback here.

**Comment by pdehaan** [comment]
^^^ OK, that was kind of funny right there, @issues-bouncer.

**Comment by ckarlof** [comment]
Too much github notification spam at the moment. We either need to get people unassigned from bugs they're not working on, or tune this a bit.

**Comment by shane-tomlinson** [comment]
> Too much github notification spam at the moment. We either need to get people unassigned from bugs they're not working on, or tune this a bit.

I found the initial round of notifications useful, though that volume of emails sent on a daily basis would be annoying. Perhaps the notifier could remember which issues it has sent notifications on, and on which date, then not issue a new reminder for that issue for another couple of weeks. Or, provide a link for devs to defer notifications for a timespan, like Bugzilla has.

**Comment by pdehaan** [comment]
I agree that the first round of messages was a bit much, but I figure it's probably just a "first time script" issue and it wouldn't be nearly as bad in subsequent runs, and typically only on Mondays since that's when we triage and assign bugs.

@eoger Does the GitHub API let us get an issue's last updated date?

What about if we spam out a message like this:

> Hey {{user}}, you have been assigned on this issue for more than 14 days!
> Maybe you should consider un-assigning yourself and let other people try to solve this issue.
> 
> _You will be automatically unassigned in 7 days._

Then, if there are no new comments/updates in 7 days, the person currently assigned to the issue is unassigned. Because the @issues-bouncer bot name-checked the person in the original reply, we'd know who was previously investigating the issue.

So basically notify on 14 days, unassign on 21 days (unless there were some comments or updates to the bug (ie: changed tags or milestones, or whatever). At which point the timer resets, and @issues-bouncer will nag in another 7 or 14 days.

**Comment by shane-tomlinson** [comment]
> So basically notify on 14 days, unassign on 21 days (unless there were some comments or updates to the bug (ie: changed tags or milestones, or whatever). At which point the timer resets, and @issues-bouncer will nag in another 7 or 14 days.

I can deal with this.

**Comment by johngruen** [comment]
@shane-tomlinson ditto

**Comment by shane-tomlinson** [comment]
We seem to have come to a general consensus. I'm closing this out.
